---
navigation:
  - function.restore-exception-handler.md: « restoreexceptionhandler
  - function.set-exception-handler.md: setexceptionhandler »
  - index.md: PHP Manual
  - ref.errorfunc.md: Функции обработки ошибок
title: seterrorhandler
---
# seterrorhandler

(PHP 4> = 4.0.1, PHP 5, PHP 7, PHP 8)

seterrorhandler — Задає користувальницький обробник помилок

### Опис

```methodsynopsis
set_error_handler(?callable $callback, int $error_levels = E_ALL): ?callable
```

Задає функцію користувача (`callback`), як обробник помилок у скрипті.

Ця функція використовується визначення власного оброблювача помилок часу виконання скрипта. Наприклад, якщо потрібно очистити дані/файли, коли відбулася критична помилка, або якщо потрібно переключити тип помилки, виходячи з якихось умов (використовуючи функцію [triggererror()](function.trigger-error.md)

Важливо пам'ятати, що стандартний обробник помилок PHP не оброблятиме жодні типи помилок, визначені в `error_levels`, поки callback-функція не поверне **`false`**. Користувальницький обробник буде викликатися за будь-якої помилки, незалежно від налаштувань, заданих функцією [errorreporting](errorfunc.configuration.md#ini.error-reporting)

Також важливо пам'ятати, що на совісті обробника лежить виклик функції [die()](function.die.md) у випадку необхідності. Якщо відбувається повернення з оброблювача помилок, керування передається наступному виразу, що стоїть за тим, що викликало помилку.

Помилки таких типів не можуть бути оброблені користувачем: **`E_ERROR`** **`E_PARSE`** **`E_CORE_ERROR`** **`E_CORE_WARNING`** **`E_COMPILE_ERROR`** **`E_COMPILE_WARNING`** незалежно від того, де вони були згенеровані і більшість помилок **`E_STRICT`**, що відбулися у файлі, де викликана функція **seterrorhandler()**

Якщо помилки виникають до запуску скрипта (наприклад, поки файл завантажується), обробний користувач не буде викликатися, якщо на цей момент він ще не зареєстрований.

### Список параметрів

`callback`

Якщо передано значення **`null`**, обробник скидається у стан за промовчанням. В іншому випадку обробник є callback-функцією з наступною сигнатурою:

```methodsynopsis
handler(    int $errno,    string $errstr,    string $errfile = ?,    int $errline = ?,    array $errcontext = ?): bool
```

`errno`

У перший аргумент `errno` буде передано рівень помилки як цілого числа.

`errstr`

У другому аргументі `errstr` буде передано повідомлення про помилку у вигляді рядка.

`errfile`

Якщо функція зворотного дзвінка приймає третій параметр `errfile`, то йому буде передано ім'я файлу, у якому сталася помилка, як рядка.

`errline`

Якщо функція зворотного дзвінка приймає четвертий параметр `errline`, то в нього буде передано номер рядка, в якому відбулася помилка, у вигляді цілого числа.

`errcontext`

Якщо функція зворотного дзвінка приймає п'ятий параметр `errcontext`, то нього буде передано масив покажчиків на активну таблицю символів у точці, де сталася помилка. Іншими словами, `errcontext` міститиме масив усіх змінних, що існують в області видимості, де сталася помилка. Користувальницькі обробники помилок не повинні змінювати цей контекст.

**Увага**

Цей параметр оголошено *Застарілим* починаючи з PHP 7.2.0 і був *ВИДАЛЕНО* у PHP 8.0.0. Якщо у вашій функції цей параметр використовується і для нього не встановлено значення за замовчуванням, то при виклику функції обробника буде видана помилка "too few arguments".

Якщо функція повертає **`false`**, керування передається вбудованому обробнику помилок.

`error_levels`

Може використовуватися для завдання маски, відповідно до якої буде викликатись `callback`, за аналогією з ini-настройкою [errorreporting](errorfunc.configuration.md#ini.error-reporting), яка відповідає за те, які помилки будуть показані у звіті. Без цієї маски `callback` буде викликатися для обробки всіх помилок, що відбуваються, незалежно від налаштувань в [errorreporting](errorfunc.configuration.md#ini.error-reporting)

### Значення, що повертаються

Повертає раніше певний обробник помилок (якщо є) Якщо на даний момент використовується вбудований обробник, функція поверне **`null`**. Якщо попередній певний обробник є методом класу, функція поверне масив, який містить ім'я класу та ім'я методу.

### список змін

| Версия | Описание |
| --- | --- |
|  | Параметр `errcontext` був видалений і більше не передається в функцію обробки помилок. |
|  | Параметр `errcontext` оголошено застарілим. Тепер при його використанні викликатиметься помилка рівня **`E_DEPRECATED`** |

### Приклади

**Приклад #1 Обробка помилок за допомогою функцій **seterrorhandler()** і [triggererror()](function.trigger-error.md)**

Приклад нижче демонструє обробку внутрішніх виключень шляхом виклику помилок різних типів та їх обробки функцією користувача:

```php
<?php
// функция обработки ошибок
function myErrorHandler($errno, $errstr, $errfile, $errline)
{
    if (!(error_reporting() & $errno)) {
        // Этот код ошибки не включён в error_reporting,
        // так что пусть обрабатываются стандартным обработчиком ошибок PHP
        return false;
    }

    // может потребоваться экранирование $errstr:
    $errstr = htmlspecialchars($errstr);

    switch ($errno) {
    case E_USER_ERROR:
        echo "<b>Пользовательская ОШИБКА</b> [$errno] $errstr<br />\n";
        echo "  Фатальная ошибка в строке $errline файла $errfile";
        echo ", PHP " . PHP_VERSION . " (" . PHP_OS . ")<br />\n";
        echo "Завершение работы...<br />\n";
        exit(1);

    case E_USER_WARNING:
        echo "<b>Пользовательское ПРЕДУПРЕЖДЕНИЕ</b> [$errno] $errstr<br />\n";
        break;

    case E_USER_NOTICE:
        echo "<b>Пользовательское УВЕДОМЛЕНИЕ</b> [$errno] $errstr<br />\n";
        break;

    default:
        echo "Неизвестная ошибка: [$errno] $errstr<br />\n";
        break;
    }

    /* Не запускаем внутренний обработчик ошибок PHP */
    return true;
}

// функция для тестирования обработчика ошибок
function scale_by_log($vect, $scale)
{
    if (!is_numeric($scale) || $scale <= 0) {
        trigger_error("log(x) для x <= 0 не определён, вы используете: scale = $scale", E_USER_ERROR);
    }

    if (!is_array($vect)) {
        trigger_error("Некорректный входной вектор, пропущен Масив значений", E_USER_WARNING);
        return null;
    }

    $temp = array();
    foreach($vect as $pos => $value) {
        if (!is_numeric($value)) {
            trigger_error("Значение на позиции $pos не является числом, будет использован 0 (ноль)", E_USER_NOTICE);
            $value = 0;
        }
        $temp[$pos] = log($scale) * $value;
    }

    return $temp;
}

// переключаемся на пользовательский обработчик
$old_error_handler = set_error_handler("myErrorHandler");

// вызовем несколько ошибок, во-первых, определим Масив с нечисловым элементом
echo "vector a\n";
$a = array(2, 3, "foo", 5.5, 43.3, 21.11);
print_r($a);

// теперь создадим ещё один Масив
echo "----\nvector b - a notice (b = log(PI) * a)\n";
/* Значение на позиции $pos не является числом, будет использован 0 (ноль)*/
$b = scale_by_log($a, M_PI);
print_r($b);

// проблема, мы передаём строку вместо Масива
echo "----\nvector c - a warning\n";
/* Некорректный входной вектор, пропущен Масив значений */
$c = scale_by_log("not array", 2.3);
var_dump($c); // NULL

// критическая ошибка, логарифм от неположительного числа не определён
echo "----\nvector d - fatal error\n";
/* log(x) для x <= 0 не определён, вы используете: scale = $scale */
$d = scale_by_log($a, -2.5);
var_dump($d); // До сюда не дойдём никогда
?>
```

Результатом виконання цього прикладу буде щось подібне:

```
vector a
Array
(
    [0] => 2
    [1] => 3
    [2] => foo
    [3] => 5.5
    [4] => 43.3
    [5] => 21.11
)
----
vector b - a notice (b = log(PI) * a)
<b>Пользовательское УВЕДОМЛЕНИЕ</b> [1024]  Значение на позиции 2 не является числом, будет использован 0 (ноль)<br />
Array
(
    [0] => 2.2894597716988
    [1] => 3.4341896575482
    [2] => 0
    [3] => 6.2960143721717
    [4] => 49.566804057279
    [5] => 24.165247890281
)
----
vector c - a warning
<b>Пользовательское ПРЕДУПРЕЖДЕНИЕ</b> [512] Некорректный входной вектор, пропущен Масив значений<br />
NULL
----
vector d - fatal error
<b>Пользовательская ОШИБКА</b> [256] log(x) for x <= 0 is undefined, you used: scale = -2.5<br />
  Фатальная ошибка в строке 35 файла trigger_error.php, PHP 5.2.1 (FreeBSD)<br />
Завершение работы...<br />
```

### Дивіться також

-   [ErrorException](class.errorexception.md)
-   [errorreporting()](function.error-reporting.md) - Задає, які помилки PHP потраплять у звіт
-   [restoreerrorhandler()](function.restore-error-handler.md) - Відновлює попередній обробник помилок
-   [triggererror()](function.trigger-error.md) - Викликає помилку користувача/попередження/повідомлення
-   [Константи рівнів помилок](errorfunc.constants.md)
